### DISCUSSION TITLE
Is there a best practice for human-in-the-loop with multiple agents ?

### DISCUSSION THREAD

**Discussion by jb-ross-bingham** [microsoft/semantic-kernel#9509] [General]
Hi, we are looking into breaking our AI Tools and behaviors into multiple agents, using the Semantic Kernel Agent Framework. However, in order to do this, the agents would need to be able to take requests from the user, interact with each other and with the user. I believe the behavior we're interested in is described in the documentation as ['Human-in-the-Loop collaboration'](https://learn.microsoft.com/en-us/semantic-kernel/frameworks/agent/?pivots=programming-language-csharp#what-problems-do-ai-agents-solve). However, I haven't seen this concept mentioned anywhere else. Is there a sample, a document, or anything else that might provide some direction for implementing this type of pattern in Semantic Kernel ?

![image](https://github.com/user-attachments/assets/91687774-3238-4e8a-ae80-73bc47119dcd)

Currently, we have an extremely long prompt that describes supported behaviors and advertises dozens of tools. This works well for interacting with the user, but it's not sustainable.

It might be worth noting that this an aspnetcore web application, where a transient instance of the Kernel is created via Http Request each time the user sends a message in the chat window. The history and state are stored, so it can be recreated with each request. When the invocation completes, a response is sent back to the user. I'm not sure if this architecture adds additional limitations to implementing a 'Human-in-the-Loop Collaboration' pattern.

I'm happy to provide some context of the problem I'm looking to solve. I started to type it, but I wasn't sure that it added anything helpful to the question.

Thank you for any information you can provide.

Note, I did do a search for similar discussion, and found one which is closed:
https://github.com/microsoft/semantic-kernel/discussions/6027

I wasn't sure if I should post a comment to a closed discussion, or open a new one...

**Comment by moonbox3** [comment]
@jb-ross-bingham thanks for your question. Have you taken a look at our Process framework? I am not sure if you're targeting SK C# or Python, but we have a getting started sample that shows a simulated "human input" step:

C# Example: https://github.com/microsoft/semantic-kernel/blob/f3633e2985e8a6bf0a9f2571c552828fec7e91fa/dotnet/samples/GettingStartedWithProcesses/Step01/Step01_Processes.cs#L89

Python Example: https://github.com/microsoft/semantic-kernel/blob/f3633e2985e8a6bf0a9f2571c552828fec7e91fa/python/samples/getting_started_with_processes/step01/step01_processes.py#L82

In C# (coming soon to Python), we then have a Processes sample around configuring steps to run the agent framework: https://github.com/microsoft/semantic-kernel/tree/main/dotnet/samples/GettingStartedWithProcesses/Step04

Would this help your use-case?

**Comment by jb-ross-bingham** [comment]
@moonbox3 , Thank you for your response, I really appreciate it. I will be looking at the Process Framework as a solution. I was aware of it, but I was hesitant to look at it in detail because of the dependency on Dapr. At least, I think I read there's a dependency on Dapr; not for local development, but for production environments.

Just to answer your question, from the limited reading I've done on the Process Framework, I believe you've nailed it, and this would solve our human interaction needs. I'm looking forward to trying it out.

Also, FWIW, we did make the switch from a no-agent ChatHistory implementation to the AgentGroupChat implementation with N number of agents. We're still building out agents, and migrating the thousands of lines of instructions into the much more compartmentalized, much more maintainable agents.

For simple scenarios, we've have early success with the human interaction. However, as the number of agents grow, and as we add increasingly complex workflow scenarios, I can see the Process Framework being even more valuable.

Thanks again for your response.
